tests/types/ember-concurrency-test.ts

```diff
@@ -0,0 +1,1883 @@
+        return false;
+      });
+
+      expect<typeof result>().toEqualTypeOf<void>();
+    }
+  });
+
+  test('forever', async () => {
+    expect<typeof forever>().toBeCallableWith();
+    expect<typeof forever>().parameters.toEqualTypeOf<[]>();
+    expect<typeof forever>().returns.toEqualTypeOf<PromiseLike<never>>();
+
+    // @ts-expect-error
+    forever('nope');
+
+    {
+      let result = await forever();
+      expect<typeof result>().toBeNever();
+    }
+  });
+});
+
+module('integration tests', () => {
+  type Resolved<T> = T extends PromiseLike<infer R> ? R : T;
+
+  test('classic ember', () => {
+    ClassicComponent.extend({
+      myTask: task(function * (immediately: boolean, ms: number = 500) {
+        if (!immediately) {
+          yield timeout(ms);
+        }
+
+        let fetchPromise = fetch('/api/data.json');
+        expect<typeof fetchPromise>().resolves.toEqualTypeOf<Response>();
+
+        let response: Response = yield fetchPromise;
+        expect<typeof response>().toEqualTypeOf<Response>();
+
+        let safeResponse: Resolved<typeof fetchPromise> = yield fetchPromise;
+        expect<typeof safeResponse>().toEqualTypeOf<Response>();
+
+        return 'wow';
+      }).restartable(),
+
+      async performMyTask() {
+        let myTask = this.get('myTask');
+
+        expect<typeof myTask>().not.toBeAny();
+        expect<typeof myTask>().toEqualTypeOf<Task<string, [boolean, number?]>>();
+        expect<typeof myTask.isRunning>().toBeBoolean();
+        expect<typeof myTask.last>().toEqualTypeOf<TaskInstance<string> | null>();
+        expect<typeof myTask.perform>().toBeCallableWith(true);
+        expect<typeof myTask.perform>().toBeCallableWith(false, 500);
+        expect<typeof myTask.perform>().parameters.toEqualTypeOf<[boolean, number?]>();
+        expect<typeof myTask.perform>().returns.toEqualTypeOf<TaskInstance<string>>();
+
+        let myTaskInstance = myTask.perform(true);
+
+        expect<typeof myTaskInstance>().not.toBeAny();
+        expect<typeof myTaskInstance>().toEqualTypeOf<TaskInstance<string>>();
+        expect<typeof myTaskInstance.isRunning>().toBeBoolean();
+        expect<typeof myTaskInstance.value>().toEqualTypeOf<string | null>();
+        expect<typeof myTaskInstance>().toMatchTypeOf<Promise<string>>();
+
+        let result = await myTaskInstance;
+
+        expect<typeof result>().not.toBeAny();
+        expect<typeof result>().toBeString();
+        expect<typeof result.length>().toBeNumber();
+
+        // @ts-expect-error
+        myTask.perform('nope');
+
+        // @ts-expect-error
+        myTask.perform(true, 'nope');
+
+        // @ts-expect-error
+        myTask.perform(false, 500, 'nope');
+      }
+    });
+  });
+
+  test('classic ember (encpsulated task)', () => {
+    ClassicComponent.extend({
+      myTask: task({
+        foo: 'foo',
+
+        *perform(immediately: boolean, ms: number = 500) {
+          let self = this;
```

Comment:
What does this binding of `this` to `self` accomplish in terms of the tests? 🤔 (No argument implied; at first glance I would *assume* it's identical, but I figure there's a reason you added it.)

Reply:
There _was_ a reason, but I don't remember what it was anymore. It has to do with `this` being a special type sometimes (hovering on it and the TSLS just shows `this: this`) but assigning it forces it to infer. But when I tired it again, it doesn't have the problem anymore so I fixed it. Maybe it was specific to TS3.6, or maybe the build will fail and I'll remember why. 🤷‍♂️